---
title: "Quality Control and Processing of Brawand et al RNA Seq data in preparation for BAGER"
author: "Erin Fry"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    theme: cerulean
    highlight: espresso
    fig_width: 8
    fig_height: 6
    self_contained: false
---

<!-- The file analysis/chunks.R contains chunks that define default settings
shared across the workflowr files. -->
```{r read-chunk, include=FALSE, cache=FALSE}
knitr::read_chunk("chunks.R")
```

<!-- Update knitr chunk options -->
```{r knitr-opts-chunk, include=FALSE}
```

<!-- Insert the date the file was last updated -->
```{r last-updated, echo=FALSE, results='asis'}
```

<!-- Insert the code version (Git commit SHA1) if Git repository exists and R
 package git2r is installed -->
```{r code-version, echo=FALSE, results='asis'}
```


# Introduction

**This R markdown file will create RNA_seq log transformed gene expression files for Bayesian Ancestral Gene Expression Reconstruction (BAGER) for six primate species and six tissues.**

 - First, the samples will be examined for successful pseudo-alignment, compared to GtEX expression data in human tissues, and scrutinized for correct density distributions. 

 - All mitochondrial genes and genes that are not expressed will then be removed before initial clustering analysis on log transformed expression values.

 - Samples that fail these criterion are then removed before initial regression analysis.
 
 - Covariates are examined for correlation with PCs and then regressed out before clustering.
 
 - A last round of sample elimination takes place before the final log-transformed processed expression files are created for each tissue.

*Special thanks Lauren Blake and [her pipeline](http://lauren-blake.github.io/Reg_Evo_Primates/analysis/).*

## Before following this pipeline

 1) Generate transcript abundance file using alignment and quantification pipeline [found here](https://github.com/erinfry6/RNASeqRealignAcrossPrimates)

 2) Obtain EnsemblID, associated gene name, and chromosomal location for genes in analysis from [Biomart](http://www.ensembl.org/biomart/martview/57d2c215b1d6b6e5cbf4662826e0d8b4): filter with gene IDs, attributes gene ID, associated gene name and chromosome, click results. Add this information into the Transcripts.txt file in path/results/qualitycontrol directory. Save as 'Transcripts.txt'

 3) Obtain [GTEX](https://www.gtexportal.org/home/) tissue expression data and save it to the results/qualitycontrol directory


```{r set directories and load libraries, include=FALSE}
## Load paths and libraries
path="/Users/lynchlab/Desktop/ErinFry/workflowr/AGER/"
pathData=paste(path,"data/",sep="")
pathResults=paste(path,"data/forBAGER/",sep="")

library(dplyr)
library(limma)
library(ggplot2)
library(gplots)
library(RColorBrewer)
library(scales)
library(edgeR)
library(R.utils)
library(plyr)
library(statmod)
library(scales)
library("ape")
```

# Sample Information and Record of Technical Variables
The following sample information was obtained directly from the authors of Brawand et al.
```{r Load data, echo=TRUE}
## read in raw expression data output of kallisto run with kmer=31
TPM_raw<-read.table(paste(pathData,"Transcript_Abundances31.txt", sep=""), header=TRUE, sep="\t", na.strings='NA', stringsAsFactors = FALSE) 

## upload gene information
geneinfo<-read.table(paste(pathData,"mart_export.txt", sep=""),header=TRUE,sep='\t',na.strings='NA',stringsAsFactors = FALSE)

## load and view the sample information
SampleInfo<-read.table(paste(pathData,"SampleInformation.txt", sep=""),sep='\t',header=TRUE)
head(SampleInfo)
## check that sample info file has the same number of samples at the TPM_raw file
nrow(SampleInfo)==ncol(TPM_raw)-1

## if there are any EnsemblIDs without a match in biomart, find their chromosomal location and gene name online
TPM_raw$Human_Ortho_EnsemblID[which(!TPM_raw$Human_Ortho_EnsemblID %in% geneinfo$Ensembl.Gene.ID)]

## combine Ensembl gene info with gene expression file
TPM_raw<-cbind(TPM_raw[,1],geneinfo[,c(2,3)],TPM_raw[,-1])
colnames(TPM_raw)<-c('EnsemblID',colnames(TPM_raw)[-1])

```

# Pre-processing of RNA-seq Data
## Expression unmapped and mapped

**Quantify the number of raw reads that mapped to genes during psuedo-alignement.**
```{r number of mapped reads}
## create list of kallisto output files
setwd(paste(pathData,"KallistoOutput/", sep=""))
listcsv<-as.character(dir(pattern="*"))# creates the list of all the csv files in the directory in true (not computer) numerical order
## check that matches sample info
length(listcsv)==nrow(SampleInfo)

## make sure samples in the correct order
SampleInfo<-SampleInfo[order(SampleInfo$SampleID),]
SampleInfo$SampleID==listcsv

## sum mapped reads and add to SampleInfo
#n equals the number of columns in SampleInfo
n=ncol(SampleInfo)

for (k in 1:nrow(SampleInfo)){ 
  SampleInfo[k,(n+1)]<-as.numeric(sum((read.delim(paste(listcsv[k],"/abundance.tsv",sep="")))$est_counts))
}
colnames(SampleInfo)<-c(colnames(SampleInfo)[-n],"Mapped_Reads")

## caclulate other mapped reads stats and add to SampleInfo
SampleInfo$Proportion_mapped_reads<-SampleInfo$Mapped_Reads/SampleInfo$Raw_reads
SampleInfo$Unmapped_reads<-SampleInfo$Raw_reads-SampleInfo$Mapped_Reads
SampleInfo$Proportion_unmapped_reads<-SampleInfo$Unmapped_reads/SampleInfo$Raw_reads

```


```{r plot mapped reads, echo=FALSE}
## plot number of mapped reads in each sample
ggplot(SampleInfo, aes(x = factor(SampleID), y = Mapped_Reads, fill = Species)) + 
  geom_bar(stat = "identity", colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  xlab("Sample") + ylab("Number of mapped reads") + 
  ggtitle("Number of Mapped reads for all samples (RNA-seq)") + 
  scale_y_continuous(labels=comma) +
  geom_hline(yintercept = 5000000)
## plot proportion of mapped reads in each sample
ggplot(SampleInfo, aes(x = factor(SampleID), y = Proportion_mapped_reads, fill = Species)) + 
  geom_bar(stat = "identity", colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  xlab("Sample") + ylab("Proportion of Mapped Reads") + 
  ggtitle("Proportion of Reads Mapped for all samples (RNA-seq)") + 
  scale_y_continuous(labels=comma)+
  geom_hline(yintercept = .2)

```

**Samples will be eliminated based on the following criteria:**

 1) if the sample fails three or more tests for quality samples it will be eliminated
 
 2) if the sample expression density curve mean is more than two standard deviations away from the mean of the other samples of that tissue's mean expression, it will be eliminated
 
 3) if sample fails to eventually cluster by species and tissue, it will be eliminated

First two tests of quality: samples that have fewer that 5 million reads or less than 20% mapped reads

```{r first failures}
## set original number of failures to 0 for each sample
SampleInfo$Failures<-0

## if does not have enough mapped reads or a low proportion, add a count
SampleInfo$Failures[which(SampleInfo$Mapped_Reads<5000000)]<-SampleInfo$Failures[which(SampleInfo$Mapped_Reads<5000000)]+1

SampleInfo$Failures[which(SampleInfo$Proportion_mapped_reads<0.2)]<-SampleInfo$Failures[which(SampleInfo$Proportion_mapped_reads<0.2)]+1

```

## GTEX Expression Comparison

To ensure our samples are from the correct tissue, check the human gene expression rank against gene expression rank in GtEX samples. Then, check non-human samples against human sample rank.

```{r format and load GTEX data, include=FALSE}
## format GTEX exression data and save, remove splice variant details, then reread into r
#GtexMedians<-read.delim("GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_median_rpkm.gct",header=F)[-c(1,2),]
#write.table(GtexMedians,"GTex_Medians_All.txt",sep='\t')
#remove excess row and column headers in excel, also get rid of versions of gene or ".3" part of gene name, replace spaces with _

GtexMedians<-read.table(paste(pathData,"Gtex_Medians_All.txt", sep=""),header=T,sep='\t')

## select tissues to compare to
GtexMedians<-select(GtexMedians, Name, Brain_._Cerebellum,Brain_._Frontal_Cortex_.BA9.,Kidney_._Cortex,Heart_._Left_Ventricle,Liver,Testis)

```

```{r only look at overlap between datasets, include=FALSE}
## Only compare genes that are present in our dataset (1-1 orthologs)
GtexMedians_inDS<- GtexMedians[GtexMedians$Name %in% TPM_raw$EnsemblID,]

GtexMedians_inDS<- filter(GtexMedians,Name %in% TPM_raw$EnsemblID)
                      
TPM_raw_inGtex<-filter(TPM_raw, EnsemblID %in% GtexMedians_inDS$Name)
nrow(TPM_raw_inGtex)==nrow(GtexMedians_inDS)


GtexMedians_inDS<-GtexMedians_inDS[order(GtexMedians_inDS$Name),]

## Examine the data files, containing gene expression in our dataset in multiple tissues and species, and the GTEX dataset with human expression levels in tissues

## transcript abundances in our data
head(TPM_raw_inGtex)

## median expression level of all genes in GTEX
head(GtexMedians_inDS)
```

```{r median expression of genes in species tissue combos, include=FALSE}
#### Calculate the median expression of each gene in each species and tissue

## specify the species and tissues in your dataset
species<-c("ggo","hsa","mml","ppa","ppy","ptr")
tissue<-c("br","cb","ht","kd","lv","ts")
tissuefull<-c("Brain","Cerebellum","Heart","Kidney","Liver","Testis")

## create labels for the samples
labels <- paste(SampleInfo$Species, SampleInfo$Tissue, sep=" ")

## create a matrix to contain average epxression levels for each species tissue combo
sampleaverages<-matrix(nrow=nrow(TPM_raw_inGtex),ncol=length(species)*length(tissue))

## populate the matrix with medians from my samples, if there is only 1, just copy over that data to this matrix
c=0
for (s in 1:length(species)){
  for (t in 1:length(tissue)){
    c<-c+1
    if(length(grep(paste(species[s],tissue[t],sep="_"), colnames(TPM_raw_inGtex)))!=1){
      
sampleaverages[,c]<-(apply(TPM_raw_inGtex[,grep(paste(species[s],tissue[t],sep="_"), colnames(TPM_raw_inGtex))],1,median)) 
} else if (length(grep(paste(species[s],tissue[t],sep="_"), colnames(TPM_raw_inGtex)))==1) {
sampleaverages[,c]<-TPM_raw_inGtex[,grep(paste(species[s],tissue[t],sep="_"), colnames(TPM_raw_inGtex))] }
  }
}

## name the columns
c=0
speciestissuecombo<-vector()
for (s in 1:length(species)){
  for (t in 1:length(tissue)){
    c<-c+1
 speciestissuecombo[c]<-paste(species[s],tissue[t],sep="_")
  }}
colnames(sampleaverages)<- speciestissuecombo

## combine with gene information
Dataaverages<-cbind(select(TPM_raw_inGtex,EnsemblID),sampleaverages)

head(Dataaverages)
```

### Compare Gene Expression Ranks between human samples and GTEX

```{r Examine Gene Expression Rank compared to GTEX, echo=FALSE}

## determine the gene expression rank in GTEX
## create new df with ranks instead of expression values
GtexMedians_inDS_rank<-cbind(select(TPM_raw_inGtex, EnsemblID),apply(GtexMedians_inDS[,-1],2,rank))

## determine the gene expression rank in my dataset
## create new df with ranks instead of expression values
TPM_raw_inGtex_rank<-cbind(select(TPM_raw_inGtex, EnsemblID),apply(Dataaverages[,-1],2,rank))

## define same tissue names in the two datasets
matchnames<-cbind(colnames(GtexMedians_inDS_rank[,-1]), c("hsa_cb","hsa_br","hsa_kd","hsa_ht","hsa_lv","hsa_ts"))

## define function for comparing rank correlations
rankcorr<-function(matchnames){
  for (i in 1:nrow(matchnames)){
  temp<-cbind(select(GtexMedians_inDS_rank, one_of(matchnames[i,1], 'tcount')), select(TPM_raw_inGtex_rank,one_of(matchnames[i,2], 'tcount')))

plot(temp[,1], temp[,2], xlab=paste("GTEX Rank", matchnames[i,1], sep=" "), ylab=paste("Dataset Rank", matchnames[i,2], sep=" "), main=paste(matchnames[i,1],"Rank Comparison, Correlation",cor.test(temp[,1],temp[,2])$estimate, sep=" "))
  }}

## a trick causes a warning, but you can ignore it and run the function for all desired tissues listed in 'matchnames'
suppressWarnings(rankcorr(matchnames))

```

```{r nonmatched tissue test, include=FALSE}
## between tissue comparisons should yield to poor correlation
## all GTex tissues against frontal cortex (br)
mismatch<-cbind(colnames(GtexMedians_inDS_rank[,-1]), c("hsa_br","hsa_cb","hsa_br","hsa_br","hsa_br","hsa_br"))
## look at correlations
suppressWarnings(rankcorr(mismatch))

```
**The correlation between human samples and GtEX is very high. This is not true when comparing two different tissue types (not shown).**

### Examine the top 100 most expressed genes in GTEX present in our dataset

```{r Top 100 expressed genes in my DS defined by GTEX, warning=FALSE, echo=FALSE}
## create a dataframe with the top 100 most expressed genes in GTEX for one tissue
Top100Gtex<-GtexMedians_inDS_rank %>% arrange(desc(Testis)) %>% slice(1:100)

## then create a dataframe with the information from our DS for those top 100 genes, expected to be the highest expressed in our human frontal cortex samples
TopGTEXindata<-filter(TPM_raw,EnsemblID %in% Top100Gtex$EnsemblID)[, c(1,(grep("ts", colnames(TPM_raw_inGtex))))]

## visualize the raw counts per gene for all genes or the top expressed genes
## boxplots of all and top genes for each human sample
suppressWarnings(boxplot(log(TPM_raw[,grep("ts", colnames(TPM_raw_inGtex))]),ylim=c(-5,12), main=paste("log TPM All Genes and top 100 in Gtex","Cerebellum",sep=" "), ylab = "log(TPM)", las = 2, cex.axis = .75))
boxplot(log(TopGTEXindata[,-1]), ylab = "log(TPM)", las = 2, cex.axis = .75, add=T, col="blue")

```

**Representative expression distribution of all genes and top 100 genes expressed in GtEX (blue):** All tissues tested followed this pattern, indicating the top most expressed in GtEX are highly expressed in our samples as well.

### Human v Non-Human Primates correlation in rank in GTEX

```{r human to non human, echo=F}
## create matrix to store information about to human average expression correlation
correlationstohuman<-matrix(ncol=1, nrow=nrow(SampleInfo))
rownames(correlationstohuman)<-SampleInfo$SampleID

## for each tissue and species, for each sample in that category, calcualte its correlation to the human average expression in that tissue
for (t in tissue){
  for (i in grep(paste("hsa_",t,sep=""), rownames(correlationstohuman))){
    correlationstohuman[i,1]<-cor.test(TPM_raw_inGtex[,(i+3)], Dataaverages[,which(colnames(Dataaverages)==paste("hsa_",t,sep=""))])$estimate
  }
  for (s in species[which(species!="hsa")]){
    for (i in grep(paste(s,t,sep="_"), rownames(correlationstohuman))){
    correlationstohuman[i,1]<-cor.test(TPM_raw_inGtex[,(i+3)], Dataaverages[,which(colnames(Dataaverages)==paste("hsa_",t,sep=""))])$estimate
  }
  }
}
## combine with sample information
correlationstohuman<-cbind(correlationstohuman,select(SampleInfo,SampleID, Species))

## visiulize these correlations
ggplot(as.data.frame(correlationstohuman), aes(x = factor(SampleID), y = correlationstohuman, fill = Species)) + 
  geom_bar(stat = "identity", colour = "black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  xlab("Sample") + ylab("Correlation to Human Median Expression") + 
  ggtitle("Correlation in Expression to Human Tissue") + 
  scale_y_continuous(labels=comma) +
  geom_hline(yintercept = .7)

```

```{r fail to correlate}
## keep track of failures - samples that have abnormally low correlation with humans
SampleInfo$Failures[which(correlationstohuman$correlationstohuman<.7)]<-SampleInfo$Failures[which(correlationstohuman$correlationstohuman<.7)]+1
```

# Processing RNA-seq data
## Eliminate mitochondrial genes and lowly expressed genes, log transform the data
```{r exclude mito genes, include=TRUE}
# eliminate all Mitochondrial genes
TPM_noMT<-filter(TPM_raw, Chromosome.Name !="MT")

## eliminate lowly expressed genes (with TPM<2 in all samples)
TPM_noMT_expressed<-TPM_noMT[rowSums(TPM_noMT[,4:ncol(TPM_noMT)])>2*nrow(SampleInfo),]

## log the data and prepare for PCA and regression analysis
TPM_noMT_expressed_log<-log2(TPM_noMT_expressed[,4:ncol(TPM_noMT_expressed)]+0.0000001)
rownames(TPM_noMT_expressed_log)<-TPM_noMT_expressed$EnsemblID
```

## Examine Expression Density Curves by Tissue
```{r density curves}
### load colors
pal <- c(brewer.pal(8, "Set1"), brewer.pal(8, "Dark2"))

## view expression density curves one tissue at a time
for (t in 1:length(tissue)){
plotDensities(TPM_noMT_expressed_log[,grep(tissue[t], colnames(TPM_noMT_expressed_log))],main=paste("log(TPM) Expression Density Curves for", tissue[t], sep=" "), col = pal)
  abline(v=log(2))
}

## identify samples with abnormal expression density curves
## these are samples with expression means that differ by more than two standard deviations from the other samples from those tissues

distsumm_noMT_expressed_log<-cbind(apply(TPM_noMT_expressed_log,2, mean),apply(TPM_noMT_expressed_log,2, sd))  ## caluclate the mean and sds of the curves
colnames(distsumm_noMT_expressed_log)<-c("mean", "sd")

abnorm<-vector() ## set vectors
for (t in tissue){
  mean.not<-vector()
  sd.not<-vector()
  abnormQ<-vector()
  for (i in 1:length(grep(t, colnames(TPM_noMT_expressed_log)))){
    mean.not[i]<-mean(distsumm_noMT_expressed_log[grep(t, colnames(TPM_noMT_expressed_log))[-i],1]) ## calculate the average mean and standard deviation expression in all tissues except the one being examined
    sd.not[i]<-sd(distsumm_noMT_expressed_log[grep(t, colnames(TPM_noMT_expressed_log))[-i],1])
    abnormQ[i]<-pnorm(q =distsumm_noMT_expressed_log[grep(t, colnames(TPM_noMT_expressed_log))[i],1] , mean = mean.not[i],sd = sd.not[i]) ## calculate the probability of seeing the mean of the one being examined
  }
  abnorm<-cbind(abnorm,colnames(TPM_noMT_expressed_log)[grep(t, colnames(TPM_noMT_expressed_log))][which(abnormQ<0.023 | abnormQ>0.977)]) ## identify samples that have mean expression more than two standard deviations away from the others
}

## which samples have bad expression density curves?
abnorm[1,]

## samples that have very abnormal distributions will be eliminated
## hsa_br_M_5 is also labeled 'temporal' instead of frontal cortex
SampleInfo$Failures[c(17,59,21,42)]<-SampleInfo$Failures[c(17,59,21,42)]+3

```


## Examine PCs and Heatmaps to identify samples that cluster poorly by tissue and/or species

```{r define PC functions, include=FALSE}
plot_scores <- function(pca, Covars,scores, n, m, points=F, legend=F){
  xmin <- min(scores[,n]) - (max(scores[,n]) - min(scores[,n]))*0.05
  if (legend == T){ ## let some room (35%) for a legend                                                                                                                                                 
    xmax <- max(scores[,n]) + (max(scores[,n]) - min(scores[,n]))*0.50
  }
  else {
    xmax <- max(scores[,n]) + (max(scores[,n]) - min(scores[,n]))*0.05
  }
  ymin <- min(scores[,m]) - (max(scores[,m]) - min(scores[,m]))*0.05
  ymax <- max(scores[,m]) + (max(scores[,m]) - min(scores[,m]))*0.05
  plot(scores[,n], scores[,m], xlab=paste("PC", n, ": ", round(summary(pca)$importance[2,n],3)*100, "% variance explained", sep=""), ylab=paste("PC", m, ": ", round(summary(pca)$importance[2,m],3)*100, "% variance explained", sep=""), main="PCA of Expression Data", xlim=c(xmin, xmax), ylim=c(ymin, ymax), type="n")
  if (points == F){
    text(scores[,n],scores[,m], rownames(scores), col="black", cex=1)
  }
  else {
    cols=pal[as.numeric(as.factor(Covars$Tissue))]
    pchs=as.numeric(as.factor(Covars$Species))
    points(scores[,n],scores[,m], col=cols, pch=pchs, cex=1.3)
    legend(x="topright",legend=unique(Covars$Tissue), fill =unique(cols))
    legend(x="topleft", legend= unique(Covars$Species), pch = unique(pchs))
  }
}

## make PC figures function
plot_PCs <- function(pca, Covars,firstPC,lastPC, points=F){
  pca_genes<-prcomp(pca, scale=F)
  scores<-pca_genes$x
  for (n in (firstPC-1):(lastPC-2)){
    n=n+1
    if (points == F){
      plot_scores(pca_genes, Covars,scores, n, n+1, points=F) 
    } else {
      plot_scores(pca_genes,Covars, scores, n, n+1, points=T) 
    }
  } }

## set seed
set.seed(6)


## Load color scale for Heatmap
colors <- colorRampPalette(c(brewer.pal(9, "Blues")[1],brewer.pal(9, "Blues")[9]))(100)


## make PC figures from desired first PC to last PC
## by tissue

plotPCandHeatmap<-function(RNA_data, Covars){
  for (i in 1:length(tissue)){
    plot_PCs(t(RNA_data[,grep(x=colnames(RNA_data),pattern = tissue[i])]),Covars,1,2)

    ## cluster by tissue
    cors <- cor(RNA_data[,grep(x=colnames(RNA_data),pattern = tissue[i])], method="spearman", use="pairwise.complete.obs")

    ## make the heatmap by tissue
    if (tissue[i]=="br"){
      heatmap.2( cors, scale="none",margins = c(12, 12), col=colors, trace='none', denscol="white", labCol=colnames(cors),cexCol = 0.2 + 1/log10(15), cexRow = 0.2 + 1/log10(15))
    } else {
        heatmap.2( cors, scale="none",margins = c(12, 12), col=colors, trace='none', denscol="white", labCol=colnames(cors), ColSideColors=pal[as.numeric(as.factor(filter(Covars,Tissue==tissuefull[i])$Species))], RowSideColors=pal[as.numeric(as.factor(filter(Covars,Tissue==tissuefull[i])$Species))], cexCol = 0.2 + 1/log10(15), cexRow = 0.2 + 1/log10(15))
      }
  }
  
  ## all samples
  plot_PCs(t(RNA_data),Covars,1,2, points=T)
  cors <- cor(RNA_data, method="spearman", use="pairwise.complete.obs")
  heatmap.2( cors, scale="none",margins = c(12, 12), col=colors, trace='none', denscol="white", labCol=colnames(cors), ColSideColors=pal[as.numeric(as.factor(Covars$Species))], RowSideColors=pal[as.numeric(as.factor(Covars$Species))], cexCol = 0.2 + 1/log10(15), cexRow = 0.2 + 1/log10(15))
  
}

```

```{r Plot the PCs before batch correction, echo=FALSE}
## plot PCs and heatmaps before eliminate samples
plotPCandHeatmap(TPM_noMT_expressed_log, SampleInfo)


```

## Eliminate samples before regression analysis

**Samples will be eliminated based on the following criteria:**

 1) if the sample fails three or more tests for quality samples it will be eliminated
 
 2) if the sample expression density curve mean is more than two standard deviations away from the mean of the other samples of that tissue's mean expression, it will be eliminated
 
 3) if sample fails to eventually cluster by species and tissue, it will be eliminated

First two tests of quality: samples that have fewer that 5 million reads or less than 20% mapped reads

```{r eliminate samples, include=TRUE}
## samples which failed to cluster properly in either PCA or heatmap
fails<-c(24,49,4,16,17)
SampleInfo$Failures[fails]<-SampleInfo$Failures[fails]+1

## clustering with the wrong tissue
fails<-c(53)
SampleInfo$Failures[fails]<-SampleInfo$Failures[fails]+3

## Eliminate samples from Sample Info / Covariates File
Sample_Info_First_Elim<-filter(SampleInfo, Failures<3)

## eliminate samples from RNA_seq file
TPM_noMT_expressed_log_First_Pass<-TPM_noMT_expressed_log[,which(SampleInfo$Failures<3)]

## be sure correctly eliminated by checking for the same number of samples in each file
nrow(Sample_Info_First_Elim)==ncol(TPM_noMT_expressed_log_First_Pass)

```

## Covariate analysis and Regression
### Choose covariates to include in analysis

```{r choose covariates and modify file, warnings=FALSE}
## find correlations between covariates and tissue or species to be sure to NOT include
SampleInfo_test<-SampleInfo[,-15]
corrwithTissue<-vector()
corrwithSpecies<-vector()
for (i in 1:ncol(SampleInfo_test)){
  corrwithTissue[i]<-suppressWarnings(chisq.test(SampleInfo$Tissue,SampleInfo_test[,i])$p.val)
  corrwithSpecies[i]<-suppressWarnings(chisq.test(SampleInfo$Species,SampleInfo_test[,i])$p.val)
}

## which are correlated with Tissue
colnames(SampleInfo_test[which(corrwithTissue<0.05/ncol(SampleInfo_test))])

## which with Species?
colnames(SampleInfo_test[which(corrwithSpecies<0.05/ncol(SampleInfo_test))])

## select covariates that should be included in regression analysis, either because they are informative, not correlated with tissue and species, or because we need to know if they are significant (and there were enough samples with the information)
Sample_Info_First_Pass<-select(Sample_Info_First_Elim, G, RIN, Raw_reads, Library_prepared_in)

```

### Define regression functions

```{r set functions for regression}
## Regression Analysis Function to calculate correlation between covariates and PCs
RegressionAnalysis<-function(RNA_DATA, covariates){
  # Calculate PCs: 
  sum.PC <- prcomp(t(RNA_DATA), scale=FALSE, center=TRUE)
  sum.PC_frame = sum.PC$x[,1:10]
  ## plot PC contribution
  plot(summary(sum.PC)$importance[1:3,1:min(20,ncol(summary(sum.PC)$importance))] [2,], main="Proportion Explained by Each PC", xlab="PC", ylab="Prop Variance Explained", type = "l")

  ## To see if covariates are correlated with a PC (looking at PC1-50)
  pval.pca1=matrix(ncol=ncol(covariates), nrow=ncol(sum.PC_frame))
  rownames(pval.pca1)=colnames(sum.PC_frame)
  colnames(pval.pca1)=colnames(covariates)
  ## for each covariate, check for correlation with PCs
  for(j in 1:ncol(covariates)){
    for(i in 1:10){
      data1= lm(sum.PC$x[,i]~covariates[,j])
      pval.pca1[i,j]=anova(data1)$'Pr(>F)'[1]
    }
  }

  return(pval.pca1)
}

## Function to regress out a specifed covariate
Regress_out<-function(RNA_DATA, covariates, ColumntoRegress){
    Regress_SOMETHING = matrix(nrow= nrow(RNA_DATA), ncol = ncol(RNA_DATA))
    rownames(Regress_SOMETHING) = rownames(RNA_DATA)
    colnames(Regress_SOMETHING) = colnames(RNA_DATA)
    ## create model for covariate and regress out its contribution
    for (i in 1:nrow(Regress_SOMETHING)) {
      model.i= lm(as.numeric(RNA_DATA[i,]) ~ covariates[,ColumntoRegress])
      Regress_SOMETHING[i,] = resid(model.i) + model.i$coefficients[1]
    }
    
    return(Regress_SOMETHING)
}
```

### Initial regression
```{r regression}
#### Initial Regression Analysis ####
First_Pass<-RegressionAnalysis(TPM_noMT_expressed_log_First_Pass, Sample_Info_First_Pass)
head(First_Pass,5)

## Define signifcance cutoff
sigcorr<-0.05/ncol(Sample_Info_First_Pass)

#### Second Pass ####

## Which covariates are the most correlated with top PCs?
which(head(First_Pass,5)<sigcorr)

## First regress out Library_prepared_in
TPM_noMT_expressed_log_Second_Pass<-Regress_out(TPM_noMT_expressed_log_First_Pass, Sample_Info_First_Pass, which(colnames(Sample_Info_First_Pass)=="Library_prepared_in"))

## Perform Second Regression Analysis
Second_Pass<-RegressionAnalysis(TPM_noMT_expressed_log_Second_Pass, Sample_Info_First_Pass)
head(Second_Pass,6)

#### Third Pass ####

## Which covariate are still correlated?
which(head(Second_Pass,6)<sigcorr)

#### Examine Clustering ####

## Check PCs and Heatmap
plotPCandHeatmap(TPM_noMT_expressed_log_Second_Pass, Sample_Info_First_Elim)
```

### Remove samples that still do not cluster with species or tissue and reregress
```{r remove final round}
## samples which failed to cluster properly in the PCA
## eliminating hsa_br_M_4 and hsa_br_M_1
fails<-c(13,16)
SampleInfo$Failures[fails]<-SampleInfo$Failures[fails]+3
## Remove from covariates file
Sample_Info_Second_Elim<-filter(SampleInfo, Failures<3)

## eliminate samples from RNA_seq file
TPM_noMT_expressed_log_First_Pass<-TPM_noMT_expressed_log[,which(SampleInfo$Failures<3)]

## be sure correctly eliminated by checking for the same number of samples in each file
nrow(Sample_Info_Second_Elim)==ncol(TPM_noMT_expressed_log_First_Pass)

## select samples that should be included in analysis, either because they are informative, not correlated with tissue and species, or because we need to know if they are significant (and there were enough samples with the information)
Sample_Info_Second_Pass<-select(Sample_Info_Second_Elim, G, RIN, Raw_reads, Library_prepared_in)

#### Second Regression Analysis ####
First_Pass<-RegressionAnalysis(TPM_noMT_expressed_log_First_Pass, Sample_Info_Second_Pass)
head(First_Pass,5)

## Define signifcance cutoff
sigcorr<-0.05/ncol(Sample_Info_First_Pass)

#### Second Pass ####

## Which covariates are the most correlated with top PCs?
which(head(First_Pass,5)<sigcorr)

## First regress out Library_prepared_in
## Regress out first covariate
TPM_noMT_expressed_log_Second_Pass<-Regress_out(TPM_noMT_expressed_log_First_Pass, Sample_Info_Second_Pass, which(colnames(Sample_Info_Second_Pass)=="Library_prepared_in"))

## Perform Second Regression Analysis
Second_Pass<-RegressionAnalysis(TPM_noMT_expressed_log_Second_Pass, Sample_Info_Second_Pass)
head(Second_Pass,6)

#### Third Pass ####

## Which covariate are still correlated?
which(head(Second_Pass,6)<sigcorr)

#### Examine Clustering ####

## Check PCs and Heatmap
plotPCandHeatmap(TPM_noMT_expressed_log_Second_Pass, Sample_Info_Second_Elim)

## define this processed matrix as the processed RNA-seq data
TPM_noMT_expressed_log_processed<-TPM_noMT_expressed_log_Second_Pass

## be sure correctly eliminated by checking for the same number of samples in each file
nrow(Sample_Info_Second_Elim)==ncol(TPM_noMT_expressed_log_processed)
``` 


# Save processed data as tissue expression files for BAGER
```{r save data files}
## set directory for files ready for BAGER

## first save all processed data
write.table(TPM_noMT_expressed_log_processed, file= paste(pathResults,Sys.Date(),"ProcessedExpData.txt", sep=""), sep='\t')

## save list of samples to be included in analysis
write.table(Sample_Info_Second_Elim, file = paste(pathResults,Sys.Date(),"IncludedSamples.txt"), sep='\t')

## save expression files by tissue
for (t in tissue){
  write.table(t(TPM_noMT_expressed_log_processed[,grep(t,colnames(TPM_noMT_expressed_log_processed))]) , paste(pathResults,Sys.Date(),"_",t,"_exp",".txt", sep=""), sep='\t', col.names = FALSE)
}
```


**by Erin Fry, `r Sys.Date()` 2017**
